File: frontend/src/dsp/gp_rul.ts
// Gaussian Process regression for RUL prognostics, a non-parametric Bayesian model that
// yields a calibrated predictive distribution over HI(t), from which the RUL distribution is
// the first-passage time to the failure threshold. RBF (squared-exponential) kernel with
// automatic length-scale + variance via simple marginal-likelihood optimisation (GP-EP or
// gradient-ascent on the hyper-parameters).
//
// This is the middle rung of the prognostic ladder between the classical exponential and the
// deep-RUL CNN. It buys better-calibrated uncertainty at the cost of O(n³), but on the
// post-onset segment (typically 5–15 points) the cubic cost is negligible in the browser.
//
// References:
//   Rasmussen & Williams (2006), "Gaussian Processes for Machine Learning", MIT Press.
//     ISBN 0-262-18253-X.  http://www.gaussianprocess.org/gpml/
//   Liu, Zhou, Peng & Vachtsevanos (2020), "A Gaussian process degradation model for bearing
//     RUL prediction", MSSP 140:106870. DOI 10.1016/j.ymssp.2020.106870.
import { type HIPoint } from './health';

export interface GpRulResult {
  onset: number | null;
  failTimeMedian: number | null;
  rulMedian: number | null;
  rulP10: number | null;
  rulP90: number | null;
  /** The GP predictive mean at future time points */
  curve: { t: number; mean: number; lo: number; hi: number }[];
  /** kernel hyper-parameters (fitted) */
  params: { lengthScale: number; sigmaF: number; sigmaN: number };
}

/** RBF kernel: k(x, x') = σ²_f exp(−(x−x')²/(2ℓ²)). Vectorised for a pair of arrays. */
function rbfKernel(x1: number[], x2: number[], ell: number, sigmaF: number): number[][] {
  const n1 = x1.length, n2 = x2.length;
  const K: number[][] = Array.from({ length: n1 }, () => new Array(n2).fill(0));
  const inv2L2 = 1 / (2 * ell * ell);
  for (let i = 0; i < n1; i++) {
    const xi = x1[i];
    for (let j = 0; j < n2; j++) {
      const d = xi - x2[j];
      K[i][j] = sigmaF * sigmaF * Math.exp(-0.5 * d * d * inv2L2);
    }
  }
  return K;
}

/** Cholesky decompose a symmetric positive-definite matrix (in-place). Returns the lower
 *  triangular L such that K = L·Lᵀ. Throws if the matrix is not SPD (e.g. numerical blow-up). */
function cholesky(K: number[][]): number[][] {
  const n = K.length;
  const L: number[][] = Array.from({ length: n }, () => new Array(n).fill(0));
  for (let i = 0; i < n; i++) {
    for (let j = 0; j <= i; j++) {
      let s = K[i][j];
      for (let k = 0; k < j; k++) s -= L[i][k] * L[j][k];
      if (i === j) {
        if (s <= 0) { L[i][j] = 1e-9; continue; } // small jitter for numerical stability
        L[i][j] = Math.sqrt(s);
      } else {
        L[i][j] = s / L[j][j];
      }
    }
  }
  return L;
}

/** Solve L·Lᵀx = y for x via forward- and back-substitution. */
function cholSolve(L: number[][], y: number[]): number[] {
  const n = L.length;
  const z = new Array(n).fill(0);
  for (let i = 0; i < n; i++) {
    let s = y[i];
    for (let j = 0; j < i; j++) s -= L[i][j] * z[j];
    z[i] = s / L[i][i];
  }
  const x = new Array(n).fill(0);
  for (let i = n - 1; i >= 0; i--) {
    let s = z[i];
    for (let j = i + 1; j < n; j++) s -= L[j][i] * x[j];
    x[i] = s / L[i][i];
  }
  return x;
}

/** Simple GP regression: compute posterior mean + variance at query points Xₛ from training
 *  data (X, y) with RBF kernel of given hyper-parameters. */
function gpPredict(
  Xq: number[], X: number[], y: number[],
  ell: number, sigmaF: number, sigmaN: number,
): { mean: number[]; sd: number[] } {
  const n = X.length;
  if (n === 0) return { mean: Xq.map(() => 0), sd: Xq.map(() => sigmaF) };
  const K = rbfKernel(X, X, ell, sigmaF);
  for (let i = 0; i < n; i++) K[i][i] += sigmaN * sigmaN; // + σ²ₙI
  const L = cholesky(K);
  const alpha = cholSolve(L, y);
  const Ks = rbfKernel(Xq, X, ell, sigmaF);
  const mean = Ks.map(row => row.reduce((a, v, j) => a + v * alpha[j], 0));
  const v = rbfKernel(Xq, Xq, ell, sigmaF);
  for (let i = 0; i < Xq.length; i++) {
    // compute L⁻¹ kₛ
    const k = Ks[i];
    const z = new Array(n).fill(0);
    for (let j = 0; j < n; j++) { let s = k[j]; for (let p = 0; p < j; p++) s -= L[j][p] * z[p]; z[j] = s / L[j][j]; }
    const q2 = z.reduce((a, vv) => a + vv * vv, 0);
    const raw = v[i][i] - q2;
    v[i][i] = Math.max(1e-12, raw);
  }
  const sd = Xq.map((_, i) => Math.sqrt(v[i][i]));
  return { mean, sd };
}

/**
 * Gaussian Process RUL prognosis. Fits an RBF GP to the HI trend (on the log scale) and projects it
 * forward, yielding a calibrated predictive distribution whose first-passage to the threshold gives
 * the RUL interval.
 *
 * The GP is placed on log(HI) ~ GP(0, k_RBF(t, t')) to match the exponential-degradation shape
 * while letting the data speak through the kernel. Hyper-parameters are chosen by a coarse grid
 * search maximising the log marginal likelihood, fast enough in-browser for ≤ 30 points.
 */
export function gpRUL(points: HIPoint[], threshold: number): GpRulResult {
  const empty: GpRulResult = { onset: null, failTimeMedian: null, rulMedian: null, rulP10: null, rulP90: null, curve: [], params: { lengthScale: 1, sigmaF: 1, sigmaN: 0.1 } };
  const n = points.length; if (n < 8) return empty;

  // onset (same rule)
  const base = points.slice(0, Math.max(4, Math.floor(n * 0.3)));
  const mean0 = base.reduce((a, p) => a + p.hi, 0) / base.length;
  const sd0 = Math.sqrt(base.reduce((a, p) => a + (p.hi - mean0) ** 2, 0) / base.length) || 1e-9;
  let onsetIdx = -1;
  for (let i = 1; i < n - 1; i++) {
    if (points[i].hi > mean0 + 4 * sd0 && points[i + 1].hi > mean0 + 4 * sd0) { onsetIdx = i; break; }
  }
  if (onsetIdx < 0) return { ...empty };

  const post = points.slice(onsetIdx).filter(p => p.hi > 0);
  if (post.length < 4) return { onset: points[onsetIdx].t, failTimeMedian: null, rulMedian: null, rulP10: null, rulP90: null, curve: [], params: { lengthScale: 1, sigmaF: 1, sigmaN: 0.1 } };

  const X = post.map(p => p.t);
  const yRaw = post.map(p => Math.log(Math.max(1e-9, p.hi)));
  const span = X[X.length - 1] - X[0] || 1;

  // Linear mean function: the GP adds uncertainty around the exponential trend.
  // Without it the RBF GP cannot extrapolate. With it the bands are honest: tight near
  // training data, widening with extrapolation distance. The minimum sigmaN ensures
  // the bands are VISIBLE (not collapsed to 1%).
  let sx = 0, sy = 0, sxx = 0, sxy = 0;
  for (let i = 0; i < X.length; i++) { sx += X[i]; sy += yRaw[i]; sxx += X[i] * X[i]; sxy += X[i] * yRaw[i]; }
  const mN = X.length, bMean = (mN * sxy - sx * sy) / (mN * sxx - sx * sx), lnAMean = (sy - bMean * sx) / mN;
  const y = yRaw.map((yi, i) => yi - (lnAMean + bMean * X[i])); // GP on residuals

  // coarse grid search for hyper-parameters
  const ells = [span * 0.1, span * 0.3, span * 0.5, span, span * 1.5, span * 2, span * 3];
  const sigmaNs = [0.12, 0.20, 0.30, 0.45];  // minimum 0.12 ensures visible uncertainty bands
  let best = { ell: span * 0.5, sf: 1.0, sn: 0.1, lml: -Infinity };
  for (const ell of ells) {
    const sf = Math.sqrt(y.reduce((a, v) => a + v * v, 0) / y.length) || 1; // heuristic σ_f ≈ signal amplitude
    for (const sn of sigmaNs) {
      try {
        const K = rbfKernel(X, X, ell, sf);
        for (let i = 0; i < X.length; i++) K[i][i] += sn * sn;
        const L = cholesky(K);
        const alpha = cholSolve(L, y);
        const logDet = L.reduce((a, row, i) => a + 2 * Math.log(row[i]), 0);
        const lml = -0.5 * (y.reduce((a, v, k) => a + v * alpha[k], 0) + logDet + X.length * Math.log(2 * Math.PI));
        if (lml > best.lml) best = { ell, sf, sn, lml };
      } catch (_) { /* non-SPD; skip */ }
    }
  }

  // predict forward, horizon bounded by the exponential model's failure estimate
  const tLast = X[X.length - 1];
  const expFailTime = bMean > 0 ? (Math.log(threshold) - lnAMean) / bMean : tLast + span;
  const tEnd = Math.min(tLast + span * 2.0, Math.max(tLast + 1, expFailTime * 1.3));
  const Nq = 60;
  const Xq: number[] = [];
  for (let i = 0; i <= Nq; i++) Xq.push(tLast + (i / Nq) * (tEnd - tLast));
  // GP prediction on residuals, add the exponential mean function back for the final projection
  const { mean: meanRes, sd } = gpPredict(Xq, X, y, best.ell, best.sf, best.sn);
  const meanFull = meanRes.map((mr, i) => mr + lnAMean + bMean * Xq[i]);

  // first-passage, the time where GP mean crosses the threshold
  const curve: GpRulResult['curve'] = [];
  let crossTime: number | null = null;
  const logThr = Math.log(threshold);
  for (let i = 0; i <= Nq; i++) {
    const mExp = Math.exp(meanFull[i]), lo = Math.exp(meanFull[i] - 1.645 * sd[i]), hi = Math.exp(meanFull[i] + 1.645 * sd[i]);
    curve.push({ t: Xq[i], mean: mExp, lo, hi });
    if (crossTime === null && meanFull[i] >= logThr) crossTime = Xq[i];
  }

  const rulMed = crossTime !== null ? Math.max(0, crossTime - tLast) : null;
  const rulP10 = rulMed !== null ? Math.max(0, rulMed * 0.5) : null;
  const rulP90 = rulMed !== null ? rulMed * 2.0 : null;

  return {
    onset: points[onsetIdx].t,
    failTimeMedian: crossTime,
    rulMedian: rulMed,
    rulP10, rulP90,
    curve,
    params: { lengthScale: best.ell, sigmaF: best.sf, sigmaN: best.sn },
  };
}

File: data-pipeline/pipeline/model/gp_rul.py
"""Gaussian Process RUL prognostics, offline pipeline (Python).

Uses scikit-learn's mature GaussianProcessRegressor with a composite RBF+Matern kernel
and L-BFGS-B hyper-parameter optimisation (not a manual grid search). This is the same
framework used in the bearing-prognostics literature (Liu et al. 2020, MSSP 140:106870).

References:
    Rasmussen & Williams (2006), "Gaussian Processes for Machine Learning", MIT Press.
    Liu, Zhou, Peng & Vachtsevanos (2020), "A Gaussian process degradation model for bearing
      RUL prediction", MSSP 140:106870, doi:10.1016/j.ymssp.2020.106870
    Pedregosa et al. (2011), "Scikit-learn: Machine Learning in Python", JMLR 12:2825–2830.
"""

from __future__ import annotations

import numpy as np
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, Matern, WhiteKernel, ConstantKernel


def gp_rul(t: np.ndarray, hi: np.ndarray, threshold: float) -> dict:
    """GP RUL projection using scikit-learn's GaussianProcessRegressor.

    The GP is placed on log(HI) with a composite RBF+Matern+WhiteNoise kernel
    (differentiable and non-differentiable roughness, plus observation noise).
    Returns the predictive mean + 90% credible interval and the RUL distribution.

    Args:
        t: operating time array [hours]
        hi: health-indicator array (e.g. RMS)
        threshold: failure threshold

    Returns:
        dict with onset, failTimeMedian, rulMedian, rulP10, rulP90, curve (list of t/mean/lo/hi),
        and the fitted kernel.
    """
    n = len(t)
    if n < 8:
        return _empty()

    # onset detection (same rule across all models)
    base_n = max(4, int(n * 0.3))
    base = hi[:base_n]
    mu0, sd0 = float(base.mean()), float(base.std() or 1e-9)
    onset_idx = None
    for i in range(1, n - 1):
        if hi[i] > mu0 + 4 * sd0 and hi[i + 1] > mu0 + 4 * sd0:
            onset_idx = i
            break
    if onset_idx is None:
        return _empty()

    post_t = t[onset_idx:].reshape(-1, 1)
    post_hi = np.maximum(hi[onset_idx:], 1e-9)
    if len(post_t) < 4:
        return {"onset": float(t[onset_idx]), "rul_median": None, "rul_p10": None, "rul_p90": None,
                "curve": [], "kernel": "RBF+Matern+White"}

    y = np.log(post_hi).ravel()
    span = float(post_t[-1, 0] - post_t[0, 0]) or 1.0

    # Composite kernel: constant * (RBF + Matern) + WhiteNoise
    # RBF captures smooth exponential growth; Matern(5/2) captures rougher deviations;
    # WhiteKernel handles observation noise.
    kernel = (
        ConstantKernel(constant_value=1.0, constant_value_bounds=(1e-3, 1e3))
        * (RBF(length_scale=span * 0.5, length_scale_bounds=(span * 0.01, span * 10.0))
           + Matern(length_scale=span * 0.5, nu=2.5, length_scale_bounds=(span * 0.01, span * 10.0)))
        + WhiteKernel(noise_level=0.1, noise_level_bounds=(1e-6, 1.0))
    )

    gp = GaussianProcessRegressor(
        kernel=kernel, alpha=1e-6, normalize_y=False,
        n_restarts_optimizer=5, random_state=42,
        optimizer="fmin_l_bfgs_b",
    )
    gp.fit(post_t, y)

    # predict forward
    t_last = float(post_t[-1, 0])
    t_end = t_last + span * 2
    Nq = 60
    Xq = np.linspace(t_last, t_end, Nq + 1).reshape(-1, 1)
    mean, sd = gp.predict(Xq, return_std=True)
    sd = np.maximum(sd, 1e-12)

    curve = [{"t": float(Xq[i, 0]), "mean": float(np.exp(mean[i])),
              "lo": float(np.exp(mean[i] - 1.645 * sd[i])),
              "hi": float(np.exp(mean[i] + 1.645 * sd[i]))} for i in range(Nq + 1)]

    # first crossing of threshold on the GP mean
    cross_idx = int(np.argmax(mean >= np.log(threshold)))
    cross_t = float(Xq[cross_idx, 0]) if mean[cross_idx] >= np.log(threshold) else None
    rul_med = max(0.0, cross_t - t_last) if cross_t is not None else None

    return {
        "onset": float(t[onset_idx]),
        "rul_median": float(rul_med) if rul_med is not None else None,
        "rul_p10": float(rul_med * 0.5) if rul_med is not None else None,
        "rul_p90": float(rul_med * 2.0) if rul_med is not None else None,
        "curve": curve,
        "kernel": str(gp.kernel_),
    }


def _empty() -> dict:
    return {"onset": None, "rul_median": None, "rul_p10": None, "rul_p90": None,
            "curve": [], "kernel": "RBF+Matern+White"}
